src/ray/raylet/scheduling_queue.h

```diff
@@ -37,6 +37,118 @@ enum class TaskState {
+  ///  be a nullptr, in which case nothing is appended.
+  /// \return Whether the removal succeeds.
+  bool RemoveTask(const TaskID &task_id,
+                  std::vector<Task> *removed_tasks = nullptr);
+
+  /// \brief Check if the queue contains a specific task id.
+  ///
+  /// \param task_id The task ID for the task.
+  /// \return Whether the task_id exists in this queue.
+  bool HasTask(const TaskID &task_id) const;
+
+  /// \brief Remove the task list of the queue.
+  /// \return A list of tasks contained in this queue.
+  const std::list<Task> &GetTasks() const;
+
+  /// \brief Recompute the resources of the tasks with minimum requirements
+  /// in the ready queue.
+  void RecomputeMinResources();
+
+protected:
+ /// A list of tasks.
+ std::list<Task> task_list_;
+ /// A hash to speed up looking up a task.
+ std::unordered_map<TaskID, std::list<Task>::iterator> task_map_;
+};
+
+class ReadyQueue : public TaskQueue {
+public:
+  /// \brief Append a task to the ready queue.
+  ///
+  /// \param task_id The task ID for the task to append.
+  /// \param task The task to append to the queue.
+  /// \return Whether the append operation succeeds.
+  bool AppendTask(const TaskID &task_id, const Task &task);
+
+  /// \brief Remove a task from the ready queue.
+  ///
+  /// \param task_id The task ID for the task to remove from the queue.
+  /// \param removed_tasks If the task specified by task_id is successfully
+  ///  removed from the queue, the task data is appended to the vector. Can
+  ///  be a nullptr, in which case nothing is appended.
+  /// \return Whether the removal succeeds.
+  bool RemoveTask(const TaskID &task_id,
+                  std::vector<Task> *removed_tasks = nullptr);
+
+  /// \brief Check whether any of the tasks in the ready queue can be scheduled.
+  ///
+  /// \param local_available_resources: resources available on local node.
+  /// \return True if there is at least a task that can be scheduled, i.e.
+  /// whose requirements can be satisfied by the local node; false otherwise.
+  const bool CanScheduleMinTask(const ResourceIdSet &local_available_resources) const;
+
+  const ResourceSet &GetCurrentResourceLoad() const { return current_resource_load_; }
+
+  /// \brief Get resources of the task with the minimum resource requirements
+  /// in the ready queue.
+  ///
+  /// \return: Minimum resources required by any task in the ready queue.
+  const ResourceSet &GetMinTaskResources() const { return min_task_resources_;};
+
+  /// \brief Get number of tasks with minimum resource requirements in the
+  /// ready queue.
+  ///
+  /// \return Number of tasks with minimum resource requirements.
+  const int &GetMinTaskCount() const { return min_task_count_;};
+
+  /// \brief Update the min  resources required by a task in the ready queue,
+  ///  when a new task is added to the queue.
+  ///
+  /// \param resources: Resource requirements of the new task being added.
+  void UpdateMinOnAdd(const ResourceSet &resources);
```

Comment:
`UpdateMinOnAdd` and `UpdateMinOnRemove` seem like private methods to me.

Reply:
Yeah, I did that first, but they are used in the tests. There is a way to make the test a friend, but that might be overkill (if it's easy to do I can do it).

Reply:
I think it is easy and preferable to make the test class a friend.

Reply:
I agree. We should keep interfaces as lean and orthogonal as possible.